src/inc/til/small_vector.h

```diff
@@ -0,0 +1,857 @@
+            std::destroy(end_new, end_old);
+            _size = end_new - beg;
+        }
+
+    private:
+        [[noreturn]] static void _throw_invalid_subscript()
+        {
+            throw std::out_of_range("invalid small_vector subscript");
+        }
+
+        [[noreturn]] static void _throw_too_long()
+        {
+            throw std::length_error("small_vector too long");
+        }
+
+        static T* _allocate(size_t size)
+        {
+            if constexpr (alignof(T) <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+            {
+                return static_cast<T*>(::operator new(size * sizeof(T)));
+            }
+            else
+            {
+                return static_cast<T*>(::operator new(size * sizeof(T), static_cast<std::align_val_t>(alignof(T))));
+            }
+        }
+
+        static void _deallocate(T* data) noexcept
+        {
+            if constexpr (alignof(T) <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+            {
+                ::operator delete(data);
+            }
+            else
+            {
+                ::operator delete(data, static_cast<std::align_val_t>(alignof(T)));
+            }
+        }
+
+        iterator _uninitialized_begin() const noexcept
+        {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+            return { _data, _data + _size, _data + _capacity };
+#else
+            return { _data };
+#endif
+        }
+
+        size_type _calculate_new_capacity(size_type min_cap) const
+        {
+            const auto new_cap = std::max(min_cap, _capacity + _capacity / 2);
+            // min_cap < _capacity indicates a overflow in the calling code.
+            // new_cap > max_size() indicates that our multiplication below would overflow.
+            if (min_cap < _capacity || new_cap > max_size())
+            {
+                _throw_too_long();
+            }
+            return new_cap;
+        }
+
+        size_type _ensure_capacity(size_type cap)
+        {
+            if (cap > _capacity) [[unlikely]]
+            {
+                _grow(cap);
+            }
+            return cap;
+        }
+
+        void _grow(size_type min_cap)
+        {
+            const auto new_cap = _calculate_new_capacity(min_cap);
+            const auto data = _allocate(new_cap);
+
+            // The earlier static_assert(std::is_nothrow_move_constructible_v<T>)
+            // ensures that we don't leak `data` here since no exceptions will be thrown.
+            std::uninitialized_move(begin(), end(), data);
+            std::destroy(begin(), end());
+
+            if (_capacity != N)
+            {
+                _deallocate(_data);
+            }
+
+            _data = data;
+            _capacity = new_cap;
+        }
+
+        template<typename F>
+        void _generic_resize(size_type new_size, F&& func)
+        {
+            if (new_size < _size)
+            {
+                std::destroy(begin() + new_size, end() - new_size);

```

Comment:
Is this right? I would have thought the second parameter should just be `end()`, or the call should be `std::destroy_n` if it's meant to be a count?

Reply:
Ah thanks! I actually did use `std::destroy_n` (and other `_n` variants) with `_size - new_size` here, and later on I replaced all those `<algorithm>` functions over to the ones without `_n` for consistency. I clearly made a bit of a mistake here - I'll have to check all the other calls as well.

Reply:
I checked all other `std::` calls and they seemed fine to me. But as they say, one is unlikely to find their own bugs, so I hope I _really_ didn't miss anything.

Thank you so much for reading the code btw!